ProjectSailor/Source/ProjectSailor/InteractionComponent.cpp

```diff
@@ -0,0 +1,185 @@
+	// ...
+	
+}
+
+
+// Called every frame
+void UInteractionComponent::TickComponent(float DeltaTime, ELevelTick TickType, FActorComponentTickFunction* ThisTickFunction)
+{
+	Super::TickComponent(DeltaTime, TickType, ThisTickFunction);
+	//calls all the time to perform raycast
+	
+	// ...
+}
+
+
+
+
+void UInteractionComponent::PerformRaycast()
+{
+	//gets a reference of player
+	APlayerController* PlayerController = Cast<APlayerController>(GetOwner()->GetInstigatorController());
+
+	// //gets a reference of Player proyecto
+	auto* player = Cast<AProjectSailorCharacter>(UGameplayStatics::GetPlayerController(GetWorld(), 0)->GetPawn());
+
+	//gets a reference of
+	if (!PlayerController) return;
+
+	//if the player exists we make a reference of cameraLocation and Rotation
+	FVector CameraLocation;
+	FRotator CameraRotation;
+
+	//returns the point of the view of the player on the variables cameraLocation and rotation
+	PlayerController->GetPlayerViewPoint(CameraLocation, CameraRotation);
+
+	// Add a vertical offset to the camera location
+	CameraLocation.Z += 50.f;
+
+	//calculate the end of the raycast and makes it unitary
+	FVector RaycastEnd = CameraLocation + CameraRotation.Vector() * 3000;
+
+	//create hitResult
+	FHitResult HitResult;
+
+	//creates the variable
+	FCollisionQueryParams CollisionParams;
+	CollisionParams.AddIgnoredActor(player);
+
+	// Definir el tipo de traza como ECC_Pawn directamente
+	ECollisionChannel TraceChannel = ECC_Pawn;
+
+	// Convertir el tipo de traza a ETraceTypeQuery
+	ETraceTypeQuery TraceTypeQuery1 = UEngineTypes::ConvertToTraceType(TraceChannel);
+
+	// Realizar el trazado de línea con el tipo de traza y los parámetros de colisión
+	bool bHit = UKismetSystemLibrary::LineTraceSingle(World, CameraLocation, RaycastEnd, TraceTypeQuery1, true, { player }, EDrawDebugTrace::ForDuration, HitResult, true);
+
+	//si ha golpeado el raycast
+	if (bHit)
+	{
+		//vemos el actor al que ha golpeado el raycast
+		AActor* HitActor = HitResult.GetActor();
+
+		UDialogueComponentNPC* DialogueComponent = Cast<UDialogueComponentNPC>(HitActor->GetComponentByClass(UDialogueComponentNPC::StaticClass()));
+		if (DialogueComponent)
+		{
+			// widget->SetVisibility(ESlateVisibility::Visible);
+			// control if the key E is pressed
+			if (pressedE)
+			{
+				//interact with HitResult
+				InteractNPC(DialogueComponent);
+				pressedE = false;
+			}
+		}
+
+		UObjectInteraction* object = Cast<UObjectInteraction>(HitActor->GetComponentByClass(UObjectInteraction::StaticClass()));
+		if (object)
+		{
+			// widget->SetVisibility(ESlateVisibility::Visible);
+			// control if the key E is pressed
+			if (pressedE)
+			{
+				//interact with HitResult
+				InteractObject(object);
+				pressedE = false;
+			}
+		}
+	}
+	else
+	{
+		// widget->SetVisibility(ESlateVisibility::Hidden);
+	}
+		
+	
+	
+	
+    
+}
+
+void UInteractionComponent::InteractNPC(UDialogueComponentNPC* dialogue)
+{
+	if (dialogue)
```

Comment:
this if block can be improved like this
`if(dialogue && dialogue->CanInteract_Implementation())`
this way we avoid using nested ifs